File: src/ctci/java/Ch1/Q6.java
package Ch1;


/**
 * Created by Michael on 3/21/2017.
 *
 * Rotate a NxN matrix 90 degrees clockwise.
 * In place and out of place
 */
public class Q6 {
    public static void main(String[] args) {
        System.out.println("Rotate NxN string matrix clockwise 90 degrees: ");
//        String[][] matrix = new String[][]{
//                {"a", "b", "c"},
//                {"d", "e", "f"},
//                {"g", "h", "i"}
//        };
        String[][] matrix = new String[][]{
                {"a", "b", "c", "z"},
                {"d", "e", "f", "k"},
                {"g", "h", "i", "l"},
                {"q", "r", "s", "t"}
        };
//        String[][] matrix = new String[][]{
//                {"a", "b"},
//                {"d", "e"}
//        };
//        String[][] matrix = new String[][]{
//                {"a"}
//        };
        System.out.println("Input: " );
        printMatrix(matrix);
        String[][] result1 = createNewRotatedMatrix(matrix);
        rotateMatrixInPlace(matrix);

        System.out.println("Create new matrix: ");
        printMatrix(result1);
        System.out.println("In place: ");
        printMatrix(matrix);
        System.out.println();

    }

    // Print out matrix in nice format
    public static void printMatrix(String[][] image) {
        for (String[] s: image) {
            for (String d: s) {
                System.out.print(d + " ");
            }
            System.out.println();
        }
    }

    // O(N^2) space, O(N^2) time
    public static String[][] createNewRotatedMatrix(String[][] image) {
        int row = image.length;
        int col = row;
        String[][] newImage = new String[row][col];

        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                newImage[j][row - 1 - i] = image[i][j];
            }
        }
        return newImage;
    }

    // In-place space, O(N^2) time
    public static void rotateMatrixInPlace(String[][] image) {
        int lastEltIndex = image.length - 1; // the N-1th index

        int thickness = image.length / 2;
        for (int i = 0; i < thickness; i++) { // completely turn a whole layer at a time
            for (int j = i; j < lastEltIndex - i; j++) { // shrinks from both sides by i
                // we are exclusive on lastEltIndex, because the first rotation in a given layer moves
                // both the first and last element of that row

                String top = image[i][j];

                // top = left
                image[i][j] = image[lastEltIndex - j][i];

                // left = bot
                image[lastEltIndex - j][i] = image[lastEltIndex - i][lastEltIndex - j];

                // bot = right
                image[lastEltIndex - i][lastEltIndex - j] = image[j][lastEltIndex - i];

                // right = top
                image[j][lastEltIndex - i] = top;
            }

        }
    }
}

File: src/ctci/python/Ch1/Q6.py
# Rotate a NxN matrix 90 degrees clockwise.
# In place and out of place

# Helper func to print matrix
def printMatrix(input):
    for x in input:
        print x

def createNewRotatedMatrix(input):
    size = len(input)
    newMatrix = [[None for x in range(size)] for y in range(size)]
    for row in range(size):
        for col in range(size):
            newMatrix[col][size - 1 - row] = input[row][col]
    return newMatrix


def rotateMatrixInPlace(input):
    lastEltIndex = len(input) - 1; # the N-1th index
    thickness = len(input) / 2
    for i in range(thickness): # completely turn a whole layer at a time
        for j in range(i, lastEltIndex - i): # shrinks from both sides by i
            # we are exclusive on lastEltIndex, because the first rotation in a given layer moves
            # both the first and last element of that row
            top = input[i][j]

            # top = left
            input[i][j] = input[lastEltIndex - j][i];

            # left = bot
            input[lastEltIndex - j][i] = input[lastEltIndex - i][lastEltIndex - j];

            # bot = right
            input[lastEltIndex - i][lastEltIndex - j] = input[j][lastEltIndex - i];

            # right = top
            input[j][lastEltIndex - i] = top;


if __name__ == '__main__':
    print "Rotate a NxN matrix 90 degrees clockwise."
    # input = [
    #     ["a", "b", "c"],
    #     ["d", "e", "f"],
    #     ["g", "h", "i"]
    # ]
    input = [
        ["a", "b", "c", "1"],
        ["d", "e", "f", "2"],
        ["g", "h", "i", "3"],
        ["j", "k", "l", "4"]
    ]
    # input = [
    #     ["a", "b"],
    #     ["d", "e"]
    # ]
    # input = [
    #     ["a"]
    # ]

    print "input: "
    printMatrix(input)

    result1 = createNewRotatedMatrix(input)
    rotateMatrixInPlace(input)
    print "create rotated: "
    printMatrix(result1)
    print "rotated in place: "
    printMatrix(input)
